File: C++/3063-Linked List Frequency.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* frequenciesOfElements(ListNode* head) {
        ListNode* tmp = head;
        unordered_map<int, int> hmap;
        
        while (tmp) {
            hmap[tmp->val]++;
            tmp = tmp->next;
        }
        
        ListNode* newHead = new ListNode();
        ListNode* prev = newHead;
        
        for (auto [_, v]:hmap) {
            newHead->next = new ListNode(v);
            newHead = newHead->next;
        }
        
        return prev->next;
    }
};

File: 3063-Linked List Frequency.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def frequenciesOfElements(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr = head;
        freqMap = defaultdict(ListNode) # value of node : ListNode(freq count, previous node)
        prev = None
        
        while curr:
            if (x:=curr.val) in freqMap:
                node = freqMap[x]
                node.val += 1
            else:
                newNode = ListNode(1, prev)
                freqMap[x] = newNode
                prev = newNode
            curr = curr.next

        return prev